### ISSUE TITLE
Deletion of several folders or files fails when using external storages.

### ISSUE THREAD

**Issue by SergioBertolinSG** [owncloud/core#28779]
### Steps to reproduce

0. Set up an external storage. For example SFTP.
1. In oC files view create three or more folders and upload files inside them.
2. Select three of the created folders and try to delete them.

### Expected behaviour
All folders are removed.

### Actual behaviour
There are errors. The first one is erased, but the next are locked.

### Server configuration

**Operating system**:
Ubuntu 16.04

**Web server:**
Apache

**Database:**
MySQL

**PHP version:**
7.0

**ownCloud version:** (see ownCloud admin page)
(current master not stable10)
{"installed":"true","maintenance":"false","needsDbUpgrade":"false","version":"10.0.3.0","versionstring":"10.0.3 beta","edition":"Community","productname":"ownCloud"}

**Updated from an older ownCloud or fresh install:**
Fresh

**The content of config/config.php:**

```

```

**Are you using external storage, if yes which one:** local/smb/sftp/...
Yes. SFTP

**Are you using encryption:** 
No.

**Logs**
```
{"reqId":"pUh0lrgWVTKkcFPHk2Y7","level":4,"time":"2017-08-23T11:00:34+00:00","remoteAddr":"HIDDEN_IP","user":"admin","app":"webdav","method":"DELETE","url":"\/remote.php\/dav\/files\/admin\/SFTP\/FOLDER3","message":"Exception: {\"Message\":\"HTTP\\\/1.1 423 \\\"SFTP\\\/FOLDER3\\\" is locked\",\"Exception\":\"OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Exception\\\\FileLocked\",\"Code\":0,\"Trace\":\"#0 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Tree.php(179): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Directory->delete()\\n#1 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(287): Sabre\\\\DAV\\\\Tree->delete('files\\\/admin\\\/SFT...')\\n#2 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpDelete(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#3 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#4 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(479): Sabre\\\\Event\\\\EventEmitter->emit('method:DELETE', Array)\\n#5 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#6 \\\/opt\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Server.php(234): Sabre\\\\DAV\\\\Server->exec()\\n#7 \\\/opt\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v2\\\/remote.php(31): OCA\\\\DAV\\\\Server->exec()\\n#8 \\\/opt\\\/owncloud\\\/remote.php(175): require_once('\\\/opt\\\/owncloud\\\/a...')\\n#9 {main}\",\"File\":\"\\\/opt\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Directory.php\",\"Line\":330,\"User\":\"admin\"}"}
{"reqId":"1VUz6STOPw52f35meTsC","level":4,"time":"2017-08-23T11:00:34+00:00","remoteAddr":"HIDDEN_IP","user":"admin","app":"webdav","method":"DELETE","url":"\/remote.php\/dav\/files\/admin\/SFTP\/FOLDER2","message":"Exception: {\"Message\":\"HTTP\\\/1.1 423 \\\"SFTP\\\/FOLDER2\\\" is locked\",\"Exception\":\"OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Exception\\\\FileLocked\",\"Code\":0,\"Trace\":\"#0 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Tree.php(179): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Directory->delete()\\n#1 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(287): Sabre\\\\DAV\\\\Tree->delete('files\\\/admin\\\/SFT...')\\n#2 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpDelete(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#3 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#4 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(479): Sabre\\\\Event\\\\EventEmitter->emit('method:DELETE', Array)\\n#5 \\\/opt\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#6 \\\/opt\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Server.php(234): Sabre\\\\DAV\\\\Server->exec()\\n#7 \\\/opt\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v2\\\/remote.php(31): OCA\\\\DAV\\\\Server->exec()\\n#8 \\\/opt\\\/owncloud\\\/remote.php(175): require_once('\\\/opt\\\/owncloud\\\/a...')\\n#9 {main}\",\"File\":\"\\\/opt\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Directory.php\",\"Line\":330,\"User\":\"admin\"}"}
```

### Client configuration

**Browser**
Chrome

Ref: https://github.com/owncloud/files_onedrive/issues/37

**Comment by PVince81** [comment]
Not happen until on stable10?

**Comment by PVince81** [comment]
Does it happen on stable10 too ?

**Comment by SergioBertolinSG** [comment]
The reference issue was tested in stable10, so it happens there as well.

**Comment by PVince81** [comment]
And likely doesn't happen on the home storage ? Does it happen with "local" as external storage ?

**Comment by SergioBertolinSG** [comment]
Happens also using local external storage. (Using master)

**Comment by PVince81** [comment]
but does not happen for a home storage ?

**Comment by PVince81** [comment]
Reproduced locally with stable10 and SFTP

**Comment by SergioBertolinSG** [comment]
> but does not happen for a home storage ?

Only happens using external storage.

**Comment by PVince81** [comment]
@SergioBertolinSG I've now tried with locking with redis and the problem did not appear.

I wonder if some deadlock or something is happening with DB based locking.

**Comment by PVince81** [comment]
With DB locking:
```
{"reqId":"K5loY0StCNZ0sCAlwQH0","level":0,"time":"2017-08-24T15:58:06+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"acquireLock: 1 \"\""}
{"reqId":"K5loY0StCNZ0sCAlwQH0","level":0,"time":"2017-08-24T15:58:06+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"acquireLock: 2 \"scanner::\""}
{"reqId":"K5loY0StCNZ0sCAlwQH0","level":0,"time":"2017-08-24T15:58:06+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"acquireLock: 1 \"\""}
{"reqId":"K5loY0StCNZ0sCAlwQH0","level":0,"time":"2017-08-24T15:58:06+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"acquireLock: 1 \"\""}
{"reqId":"tFQFFO6jW9Q9V04QqMH2","level":0,"time":"2017-08-24T15:58:06+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/c","message":"acquireLock: 1 \"\""}
{"reqId":"tFQFFO6jW9Q9V04QqMH2","level":0,"time":"2017-08-24T15:58:06+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/c","message":"acquireLock: 2 \"scanner::\""}
{"reqId":"K5loY0StCNZ0sCAlwQH0","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"releaseLock: 1 \"\""}
{"reqId":"K5loY0StCNZ0sCAlwQH0","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"acquireLock: 1 \"sub\""}
{"reqId":"K5loY0StCNZ0sCAlwQH0","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"releaseLock: 1 \"sub\""}
{"reqId":"K5loY0StCNZ0sCAlwQH0","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"acquireLock: 1 \"b\""}
{"reqId":"K5loY0StCNZ0sCAlwQH0","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"releaseLock: 1 \"b\""}
{"reqId":"K5loY0StCNZ0sCAlwQH0","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"acquireLock: 1 \"c\""}
{"reqId":"K5loY0StCNZ0sCAlwQH0","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"releaseLock: 1 \"c\""}
{"reqId":"K5loY0StCNZ0sCAlwQH0","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"acquireLock: 1 \"a\""}
{"reqId":"K5loY0StCNZ0sCAlwQH0","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"releaseLock: 1 \"a\""}
{"reqId":"K5loY0StCNZ0sCAlwQH0","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"releaseLock: 1 \"\""}
{"reqId":"K5loY0StCNZ0sCAlwQH0","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"releaseLock: 2 \"scanner::\""}
{"reqId":"K5loY0StCNZ0sCAlwQH0","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"releaseLock: 1 \"\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"acquireLock: 1 \"\""}
{"reqId":"K5loY0StCNZ0sCAlwQH0","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"acquireLock: 1 \"b\""}
{"reqId":"K5loY0StCNZ0sCAlwQH0","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"acquireLock: 1 \"\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"acquireLock: 2 \"scanner::\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"acquireLock: 1 \"\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"acquireLock: 1 \"\""}
{"reqId":"tFQFFO6jW9Q9V04QqMH2","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/c","message":"acquireLock: 1 \"c\""}
{"reqId":"tFQFFO6jW9Q9V04QqMH2","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/c","message":"acquireLock: 1 \"\""}
{"reqId":"K5loY0StCNZ0sCAlwQH0","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"changeLock: 2 \"b\""}
{"reqId":"tFQFFO6jW9Q9V04QqMH2","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/c","message":"changeLock: 2 \"c\""}
{"reqId":"tFQFFO6jW9Q9V04QqMH2","level":4,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"webdav","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/c","message":"Exception: {\"Message\":\"HTTP\\\/1.1 423 \\\"sftp\\\/c\\\" is locked\",\"Exception\":\"OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Exception\\\\FileLocked\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Tree.php(179): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Directory->delete()\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(287): Sabre\\\\DAV\\\\Tree->delete('files\\\/admin\\\/sft...')\\n#2 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpDelete(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#4 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(479): Sabre\\\\Event\\\\EventEmitter->emit('method:DELETE', Array)\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Server.php(234): Sabre\\\\DAV\\\\Server->exec()\\n#7 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v2\\\/remote.php(31): OCA\\\\DAV\\\\Server->exec()\\n#8 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/remote.php(175): require_once('\\\/srv\\\/www\\\/htdocs...')\\n#9 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Directory.php\",\"Line\":330,\"User\":\"admin\"}"}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"releaseLock: 1 \"\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"acquireLock: 1 \"sub\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"releaseLock: 1 \"sub\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"acquireLock: 1 \"b\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"acquireLock: 1 \"a\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"acquireLock: 1 \"\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"acquireLock: 2 \"scanner::a\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"acquireLock: 1 \"a\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"acquireLock: 1 \"a\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"releaseLock: 1 \"a\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"acquireLock: 1 \"a\/data1.dat\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"releaseLock: 1 \"a\/data1.dat\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"releaseLock: 1 \"a\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"releaseLock: 2 \"scanner::a\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"releaseLock: 1 \"a\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"releaseLock: 1 \"\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"acquireLock: 1 \"a\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"acquireLock: 1 \"\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":0,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"changeLock: 2 \"a\""}
{"reqId":"5znIiOYwLf2rj1YaRn18","level":4,"time":"2017-08-24T15:58:07+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"webdav","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"Exception: {\"Message\":\"HTTP\\\/1.1 423 \\\"sftp\\\/a\\\" is locked\",\"Exception\":\"OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Exception\\\\FileLocked\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Tree.php(179): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Directory->delete()\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(287): Sabre\\\\DAV\\\\Tree->delete('files\\\/admin\\\/sft...')\\n#2 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpDelete(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#4 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(479): Sabre\\\\Event\\\\EventEmitter->emit('method:DELETE', Array)\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Server.php(234): Sabre\\\\DAV\\\\Server->exec()\\n#7 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v2\\\/remote.php(31): OCA\\\\DAV\\\\Server->exec()\\n#8 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/remote.php(175): require_once('\\\/srv\\\/www\\\/htdocs...')\\n#9 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Directory.php\",\"Line\":330,\"User\":\"admin\"}"}
```

**Comment by PVince81** [comment]
There are scanner entries for the locks. I tried setting the ext storage mode to "Never" check for changes, still the locking errors appear. Well, now it's worse: I get 403 for all three calls:

```
{"reqId":"aEe3ailMDwK3g8voUwJ3","level":0,"time":"2017-08-24T16:02:56+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/c","message":"acquireLock: 1 \"c\""}
{"reqId":"aEe3ailMDwK3g8voUwJ3","level":0,"time":"2017-08-24T16:02:56+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/c","message":"acquireLock: 1 \"\""}
{"reqId":"aEe3ailMDwK3g8voUwJ3","level":0,"time":"2017-08-24T16:02:56+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/c","message":"changeLock: 2 \"c\""}
{"reqId":"YLFZfreYEFI9VAxYOXu4","level":0,"time":"2017-08-24T16:02:56+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"acquireLock: 1 \"a\""}
{"reqId":"YLFZfreYEFI9VAxYOXu4","level":0,"time":"2017-08-24T16:02:56+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"acquireLock: 1 \"\""}
{"reqId":"YLFZfreYEFI9VAxYOXu4","level":0,"time":"2017-08-24T16:02:56+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"changeLock: 2 \"a\""}
{"reqId":"6FMaUHxwrabqWDO8tTdF","level":0,"time":"2017-08-24T16:02:56+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"acquireLock: 1 \"b\""}
{"reqId":"6FMaUHxwrabqWDO8tTdF","level":0,"time":"2017-08-24T16:02:56+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"acquireLock: 1 \"\""}
{"reqId":"6FMaUHxwrabqWDO8tTdF","level":0,"time":"2017-08-24T16:02:56+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"changeLock: 2 \"b\""}
{"reqId":"aEe3ailMDwK3g8voUwJ3","level":0,"time":"2017-08-24T16:03:17+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/c","message":"changeLock: 1 \"c\""}
{"reqId":"aEe3ailMDwK3g8voUwJ3","level":0,"time":"2017-08-24T16:03:17+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/c","message":"releaseLock: 1 \"c\""}
{"reqId":"aEe3ailMDwK3g8voUwJ3","level":0,"time":"2017-08-24T16:03:17+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/c","message":"releaseLock: 1 \"\""}
{"reqId":"6FMaUHxwrabqWDO8tTdF","level":0,"time":"2017-08-24T16:03:17+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"changeLock: 1 \"b\""}
{"reqId":"YLFZfreYEFI9VAxYOXu4","level":0,"time":"2017-08-24T16:03:17+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"changeLock: 1 \"a\""}
{"reqId":"YLFZfreYEFI9VAxYOXu4","level":0,"time":"2017-08-24T16:03:17+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"releaseLock: 1 \"a\""}
{"reqId":"6FMaUHxwrabqWDO8tTdF","level":0,"time":"2017-08-24T16:03:17+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"releaseLock: 1 \"b\""}
{"reqId":"YLFZfreYEFI9VAxYOXu4","level":0,"time":"2017-08-24T16:03:17+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"releaseLock: 1 \"\""}
{"reqId":"6FMaUHxwrabqWDO8tTdF","level":0,"time":"2017-08-24T16:03:17+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"core","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"releaseLock: 1 \"\""}
{"reqId":"aEe3ailMDwK3g8voUwJ3","level":0,"time":"2017-08-24T16:03:17+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"webdav","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/c","message":"Exception: {\"Message\":\"HTTP\\\/1.1 403 Forbidden\",\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\Forbidden\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Tree.php(179): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Directory->delete()\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(287): Sabre\\\\DAV\\\\Tree->delete('files\\\/admin\\\/sft...')\\n#2 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpDelete(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#4 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(479): Sabre\\\\Event\\\\EventEmitter->emit('method:DELETE', Array)\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Server.php(234): Sabre\\\\DAV\\\\Server->exec()\\n#7 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v2\\\/remote.php(31): OCA\\\\DAV\\\\Server->exec()\\n#8 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/remote.php(175): require_once('\\\/srv\\\/www\\\/htdocs...')\\n#9 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Directory.php\",\"Line\":325,\"User\":\"admin\"}"}
{"reqId":"YLFZfreYEFI9VAxYOXu4","level":0,"time":"2017-08-24T16:03:17+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"webdav","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/a","message":"Exception: {\"Message\":\"HTTP\\\/1.1 403 Forbidden\",\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\Forbidden\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Tree.php(179): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Directory->delete()\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(287): Sabre\\\\DAV\\\\Tree->delete('files\\\/admin\\\/sft...')\\n#2 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpDelete(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#4 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(479): Sabre\\\\Event\\\\EventEmitter->emit('method:DELETE', Array)\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Server.php(234): Sabre\\\\DAV\\\\Server->exec()\\n#7 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v2\\\/remote.php(31): OCA\\\\DAV\\\\Server->exec()\\n#8 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/remote.php(175): require_once('\\\/srv\\\/www\\\/htdocs...')\\n#9 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Directory.php\",\"Line\":325,\"User\":\"admin\"}"}
{"reqId":"6FMaUHxwrabqWDO8tTdF","level":0,"time":"2017-08-24T16:03:17+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"webdav","method":"DELETE","url":"\/owncloud\/remote.php\/dav\/files\/admin\/sftp\/b","message":"Exception: {\"Message\":\"HTTP\\\/1.1 403 Forbidden\",\"Exception\":\"Sabre\\\\DAV\\\\Exception\\\\Forbidden\",\"Code\":0,\"Trace\":\"#0 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Tree.php(179): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\Directory->delete()\\n#1 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(287): Sabre\\\\DAV\\\\Tree->delete('files\\\/admin\\\/sft...')\\n#2 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpDelete(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#3 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#4 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(479): Sabre\\\\Event\\\\EventEmitter->emit('method:DELETE', Array)\\n#5 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/lib\\\/composer\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(254): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#6 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Server.php(234): Sabre\\\\DAV\\\\Server->exec()\\n#7 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/appinfo\\\/v2\\\/remote.php(31): OCA\\\\DAV\\\\Server->exec()\\n#8 \\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/remote.php(175): require_once('\\\/srv\\\/www\\\/htdocs...')\\n#9 {main}\",\"File\":\"\\\/srv\\\/www\\\/htdocs\\\/owncloud\\\/apps\\\/dav\\\/lib\\\/Connector\\\/Sabre\\\/Directory.php\",\"Line\":325,\"User\":\"admin\"}"}
```

**Comment by PVince81** [comment]
For the case with "check updates" enabled which matches this original report, what I observe when deleting three folders "a", "b" and "c" on external storage:

The DELETE call for "b" apparently also shortly sets a shared lock on "a", "b" and also "c" instead of just "b". Not sure why, could be the scanner checking their `hasUpdated` values.
If at exactly this time, the concurrent DELETE for "a" comes, it tries to get an exclusive lock on "a". If "a" has a shared lock due to the above, it will fail with "LockedException" which is expected.

Now the first thing that comes to mind is the RetryLockingWrapper attempt I made: https://github.com/owncloud/core/pull/28544 (unfinished, buggy). If this one could work then instead of failing directly it would at least retry later.

Next steps:
- confirm that the shared locks above are from the scanner
- if yes, find out why disabling the scanner (set "check for updates" to "never") still has issues
- think of alternatives to fix this aside from RetryLockingWrapper

cc @jvillafanez

**Comment by jvillafanez** [comment]
There is more research about it in https://github.com/owncloud/files_onedrive/issues/37#issuecomment-324074762

**Comment by PVince81** [comment]
Thanks, this looks scary: a caldav plugin triggering opendir / scanning ?!

In general I wondered whether we should simply disable any scanning when the operation is not PROPFIND, GET or MOVE. It doesn't make much sense to rescan for changes for a DELETE.

Hmmm, well... maybe it does, because a DELETE moves stuff to trashbin and needs the filecache entries to be up to date before moving ?

**Comment by jvillafanez** [comment]
I think the most important thing is to know what is causing the lock of the parent folder, which, I guess, is what is preventing the rest of the deletions to be executed. Ideally, you should only lock the folder you're going to delete, so you shouldn't collide with the deletion of other sibling folders.

Whether we should (deep) scan the deleted folder or not, that's a decision for the trashbin app, not for core, but it should only affect to the folder that's going to be deleted. This implies that deletion of sibling folders shouldn't collide with eachother

**Comment by SergioBertolinSG** [comment]
Tested using redis (https://github.com/owncloud/files_onedrive/issues/37#issuecomment-324879207)

Works file, no errors.

But another bug has appeared:

https://github.com/owncloud/core/issues/28800

**Comment by jvillafanez** [comment]
Did you tested with several big folders? In this case the deletion should take some time since it needs to move the contents to the trashbin.

If it works in this scenario with redis, it will likely be a problem in the DB implementation.

**Comment by PVince81** [comment]
I just tried with the RetryLockingProvider out of curiosity, but it doesn't solve the problem: https://github.com/owncloud/core/pull/28544. Anyway, this PR isn't ready as it doesn't seem to retry properly.

**Comment by PVince81** [comment]
Oh, if I set "filelocking.retrydelay" to 5000 in https://github.com/owncloud/core/pull/28544 then the deletion work. But I suspect that the retry delay might need to be high enough to cover for the rescan cost... Might be a quick general workaround but the real solution would be to prevent the rescan in the first place.

**Comment by SergioBertolinSG** [comment]
No happening using 10.0.2.

**Comment by PVince81** [comment]
Ah! Then there is hope!

**Comment by PVince81** [comment]
I'll bisect this

**Comment by PVince81** [comment]
:cry: here it is: switching the web UI to use the new DAV endpoint: 99f6b0435d78bba51043dbe14ad78069a710293c from https://github.com/owncloud/core/pull/28769

@jvillafanez mentioned that some additional plugins were registered on the new DAV endpoint and might interfere with extra locks due to extra scans. I'll see if I can prevent them to operate when on the files endpoint.

**Comment by PVince81** [comment]
In general I'm already unhappy that this massive list of Sabre plugins are registered for every call instead of just for the routes they are operating on. This is a limitaion of the Sabre plugin system and can only be worked around by adding explicit checks either in the plugin code or before registration.

@DeepDiver1975 FYI ^

**Comment by PVince81** [comment]
So I tried to disable almost all Sabre plugins on the new dav endpoint but am still getting locking errors.

There might still be other subtle differences to look into.

Another mention of @jvillafanez was that the trashbin seemed to interfere. But I don't see why it would with the new endpoint but not the old one.

**Comment by PVince81** [comment]
Switching to old endpoint on stable10 makes deletion work correctly. This tells us that there was no other change between 10.0.3beta and stable10 that would globally break this.

So the problem is indeed specific to the new dav endpoint, even with almost all plugins disabled.

**Comment by PVince81** [comment]
old endpoint:
```
0  OCA\Files_Trashbin\Trashbin::move2trash() /srv/www/htdocs/owncloud/apps/files_trashbin/lib/Trashbin.php:239
1  OCA\Files_Trashbin\Storage->doDelete() /srv/www/htdocs/owncloud/apps/files_trashbin/lib/Storage.php:185
2  OCA\Files_Trashbin\Storage->rmdir() /srv/www/htdocs/owncloud/apps/files_trashbin/lib/Storage.php:131
3  OC\Files\View->basicOperation() /srv/www/htdocs/owncloud/lib/private/Files/View.php:1123
4  OC\Files\View->rmdir() /srv/www/htdocs/owncloud/lib/private/Files/View.php:341
5  OCA\DAV\Connector\Sabre\Directory->delete() /srv/www/htdocs/owncloud/apps/dav/lib/Connector/Sabre/Directory.php:323
6  OCA\DAV\Connector\Sabre\ObjectTree->delete() /srv/www/htdocs/owncloud/lib/composer/sabre/dav/lib/DAV/Tree.php:179
7  Sabre\DAV\CorePlugin->httpDelete() /srv/www/htdocs/owncloud/lib/composer/sabre/dav/lib/DAV/CorePlugin.php:287
8  call_user_func_array:{/srv/www/htdocs/owncloud/lib/composer/sabre/event/lib/EventEmitterTrait.php:105}() /srv/www/htdocs/owncloud/lib/composer/sabre/event/lib/EventEmitterTrait.php:105
9  OCA\DAV\Connector\Sabre\Server->emit() /srv/www/htdocs/owncloud/lib/composer/sabre/event/lib/EventEmitterTrait.php:105
10 OCA\DAV\Connector\Sabre\Server->invokeMethod() /srv/www/htdocs/owncloud/lib/composer/sabre/dav/lib/DAV/Server.php:479
11 OCA\DAV\Connector\Sabre\Server->exec() /srv/www/htdocs/owncloud/lib/composer/sabre/dav/lib/DAV/Server.php:254
12 require_once()  /srv/www/htdocs/owncloud/apps/dav/appinfo/v1/webdav.php:63
13 {main}          /srv/www/htdocs/owncloud/remote.php:165
```

new endpoint:
```
0  OCA\Files_Trashbin\Trashbin::move2trash() /srv/www/htdocs/owncloud/apps/files_trashbin/lib/Trashbin.php:239
1  OCA\Files_Trashbin\Storage->doDelete() /srv/www/htdocs/owncloud/apps/files_trashbin/lib/Storage.php:185
2  OCA\Files_Trashbin\Storage->rmdir() /srv/www/htdocs/owncloud/apps/files_trashbin/lib/Storage.php:131
3  OC\Files\View->basicOperation() /srv/www/htdocs/owncloud/lib/private/Files/View.php:1123
4  OC\Files\View->rmdir() /srv/www/htdocs/owncloud/lib/private/Files/View.php:341
5  OCA\DAV\Connector\Sabre\Directory->delete() /srv/www/htdocs/owncloud/apps/dav/lib/Connector/Sabre/Directory.php:323
6  Sabre\DAV\Tree->delete() /srv/www/htdocs/owncloud/lib/composer/sabre/dav/lib/DAV/Tree.php:179
7  Sabre\DAV\CorePlugin->httpDelete() /srv/www/htdocs/owncloud/lib/composer/sabre/dav/lib/DAV/CorePlugin.php:287
8  call_user_func_array:{/srv/www/htdocs/owncloud/lib/composer/sabre/event/lib/EventEmitterTrait.php:105}() /srv/www/htdocs/owncloud/lib/composer/sabre/event/lib/EventEmitterTrait.php:105
9  OCA\DAV\Connector\Sabre\Server->emit() /srv/www/htdocs/owncloud/lib/composer/sabre/event/lib/EventEmitterTrait.php:105
10 OCA\DAV\Connector\Sabre\Server->invokeMethod() /srv/www/htdocs/owncloud/lib/composer/sabre/dav/lib/DAV/Server.php:479
11 OCA\DAV\Connector\Sabre\Server->exec() /srv/www/htdocs/owncloud/lib/composer/sabre/dav/lib/DAV/Server.php:254
12 OCA\DAV\Server->exec() /srv/www/htdocs/owncloud/apps/dav/lib/Server.php:254
13 require_once()  /srv/www/htdocs/owncloud/apps/dav/appinfo/v2/remote.php:31
14 {main}          /srv/www/htdocs/owncloud/remote.php:165
```

There's a subtle difference: old endpoint goes through `ObjectTree->delete()`. Let's continue looking there...

**Comment by PVince81** [comment]
I added some logging in the locking code to log what path is logged when, including stack traces, and this for both endpoints. There are bigger differences: https://cloud.owncloud.com/index.php/s/VK1ncpEUdvhScz9

Now need to take some time to analyze this...

**Comment by PVince81** [comment]
additional note: the above was produced with minimal Sabre plugins on new dav

**Comment by PVince81** [comment]
The first `acquireLock` in old DAV is done directly on "a", the folder to delete.
In new DAV, it acquires a lock on the parent.

It seems it's due to the differences in `getNodeForPath` from ObjectTree which we had in old dav.
We don't have it any more in new dav, so it falls back to `Tree::getNodeForPath` which itself first does a `getNodeForPath` on the parent node, which itself would cause a `getFileInfo()` on the parent, which explains why a lock is set higher. And calling `getFileInfo()` would likely trigger additional `hasUpdated` calls (to be confirmed).

I don't think we can use our custom ObjectTree like we did before, because now the node tree isn't only about files but also about collections of different types.

So we either need to make sure that such `getFileInfo` calls on the parents don't trigger additional locks or scans, or find a way to provide a shortcut for quicker node access in the file tree collections.

**Comment by PVince81** [comment]
Making FileInfo lazy for the tree traversal algo is not an option because we need to getFileInfo to find out if a node even exists.

So far the only way I see is to bring back a custom ObjectTree implementation which itself will detect if we're querying a under "files/$userId". If yes, then shortcut a bit and go directly to the target node.

**Comment by PVince81** [comment]
- [ ] bring back ObjectTree for direct node access
- [ ] calendar plugins messing up ?
- [ ] DAVACL plugin messing up

**Comment by PVince81** [comment]
Pfff, the reason why DAVACL is messing up too is because it's checking privileges on all parents recursively. So when we do an operation on "files/$userId/sftp/a" it will recurse through every parent and call `getNodeForPah()` on each, which triggers the same issue like the default object tree.

We could prevent DAVACL to check every parent when dealing with the "files/*" subtree as a workaround...

Seems these are many workarounds due to the fact that accessing every parent is expensive and messes up with locks.

**Comment by PVince81** [comment]
Another completely different solution that comes to mind would be to only allow rescanning of everything that is below the currently requested node. If one of the parent nodes is retrieved, don't rescan or check for changes. This would likely solve the locking issue but not the filecache performance issue as we'd still be querying the file cache for every parent due to Sabre node iteration.

**Comment by PVince81** [comment]
The subtle thing about parallel operations on an external storage with update detection enabled is that one of the operation operates on disk and the second operation might detect this change on-disk first before the first operation finishes. This is why we need to keep scanner locking anyway.

**Comment by PVince81** [comment]
First working fix, the one with the tree shortcut and disabling DAVACL: https://github.com/owncloud/core/pull/28853

@jvillafanez @SergioBertolinSG this solves the problem but might not be a satisfactory solution.

I'll have a look if an alternative solution like https://github.com/owncloud/core/issues/28779#issuecomment-325979900 is possible...

**Comment by PVince81** [comment]
Worst case we revert web UI to use the old DAV endpoint. However this doesn't exclude issues with parallel deletion initiated by newer desktop clients which use the new DAV endpoint...

**Comment by PVince81** [comment]
POC of alternative solution: limits the watcher to any subpaths of the currently requested path: https://github.com/owncloud/core/pull/28855. So whenever any of the parents gets queried by the Sabre tree, the watcher/scanner doesn't kick in.

The part I dislike about this approach is that it doesn't solve the parent querying / filecache queries like https://github.com/owncloud/core/pull/28853 does.

**Comment by PVince81** [comment]
Despite the fact that the desktop client 2.3.3 is using the new DAV endpoint and is doing the three DELETE operations in parallel, I didn't manage to run into the locking situation.

Maybe the browser doesn't do it exactly in parallel but with a slight microscopic delay ?

Anyway, this makes this bug less critical. I think we'll switch the web UI to old DAV in the meantime.

Because the fix itself https://github.com/owncloud/core/pull/28853 is complex and requires more discussions regarding the DAV architecture.

**Comment by PVince81** [comment]
This PR reverts the web UI back to the old DAV endpoint as a workaround for the next release: https://github.com/owncloud/core/pull/28914

**Comment by SergioBertolinSG** [comment]
This works in current master, close?

**Comment by PVince81** [comment]
Ok, let's close. We'll need to redo the switch to new DAV, I'll put this on my PR https://github.com/owncloud/core/pull/28853

**Comment by PVince81** [comment]
actually I've put the switch in a separate PR here: https://github.com/owncloud/core/pull/28874 which also contains switching for single file uploads
